File: mobile/lib/src/theme/category_visuals.dart
import 'package:flutter/material.dart';

/// A color + icon pairing used to render a category in a colorful, recognizable
/// way (the "category-rich" look of top finance apps).
class CategoryVisual {
  const CategoryVisual(this.color, this.icon);

  final Color color;
  final IconData icon;
}

/// Vibrant, evenly-spaced palette used when a category has no explicit color.
const _palette = [
  Color(0xFFEF4444), // red
  Color(0xFFF97316), // orange
  Color(0xFFF59E0B), // amber
  Color(0xFF10B981), // emerald
  Color(0xFF06B6D4), // cyan
  Color(0xFF3B82F6), // blue
  Color(0xFF6366F1), // indigo
  Color(0xFF8B5CF6), // violet
  Color(0xFFEC4899), // pink
  Color(0xFF14B8A6), // teal
];

/// Keyword → icon mapping so common categories get a fitting glyph even when
/// no icon is stored. First matching keyword wins.
const _iconKeywords = <String, IconData>{
  'food': Icons.restaurant,
  'restaurant': Icons.restaurant,
  'grocer': Icons.local_grocery_store,
  'lunch': Icons.lunch_dining,
  'dinner': Icons.dinner_dining,
  'coffee': Icons.local_cafe,
  'transport': Icons.directions_bus,
  'uber': Icons.local_taxi,
  'taxi': Icons.local_taxi,
  'fuel': Icons.local_gas_station,
  'gas': Icons.local_gas_station,
  'car': Icons.directions_car,
  'travel': Icons.flight,
  'flight': Icons.flight,
  'rent': Icons.home,
  'home': Icons.home,
  'house': Icons.house,
  'bill': Icons.receipt_long,
  'utilit': Icons.bolt,
  'electric': Icons.bolt,
  'water': Icons.water_drop,
  'internet': Icons.wifi,
  'phone': Icons.smartphone,
  'mobile': Icons.smartphone,
  'shop': Icons.shopping_bag,
  'cloth': Icons.checkroom,
  'health': Icons.favorite,
  'medic': Icons.medical_services,
  'pharma': Icons.medical_services,
  'gym': Icons.fitness_center,
  'fitness': Icons.fitness_center,
  'education': Icons.school,
  'school': Icons.school,
  'book': Icons.menu_book,
  'entertain': Icons.movie,
  'movie': Icons.movie,
  'music': Icons.music_note,
  'game': Icons.sports_esports,
  'gift': Icons.card_giftcard,
  'donation': Icons.volunteer_activism,
  'charity': Icons.volunteer_activism,
  'salary': Icons.payments,
  'income': Icons.savings,
  'invest': Icons.trending_up,
  'saving': Icons.savings,
  'tax': Icons.account_balance,
  'bank': Icons.account_balance,
  'insurance': Icons.shield,
  'pet': Icons.pets,
  'baby': Icons.child_friendly,
  'child': Icons.child_care,
  'beauty': Icons.spa,
  'salon': Icons.content_cut,
  'transfer': Icons.swap_horiz,
};

/// Resolves a category's visual identity: an explicit hex [color] if present,
/// otherwise a stable palette color derived from [name]; and an icon inferred
/// from the name's keywords, falling back to a generic category glyph.
CategoryVisual categoryVisual({String? name, String? color}) {
  final lower = (name ?? '').toLowerCase();
  final resolvedColor = _parseHexColor(color) ?? _colorForName(lower);
  return CategoryVisual(resolvedColor, _iconForName(lower));
}

Color _colorForName(String name) {
  if (name.isEmpty) return _palette[0];
  var hash = 0;
  for (final unit in name.codeUnits) {
    hash = (hash * 31 + unit) & 0x7fffffff;
  }
  return _palette[hash % _palette.length];
}

IconData _iconForName(String name) {
  for (final entry in _iconKeywords.entries) {
    if (name.contains(entry.key)) return entry.value;
  }
  return Icons.category_rounded;
}

Color? _parseHexColor(String? value) {
  if (value == null) return null;
  var hex = value.trim().replaceAll('#', '');
  if (hex.length == 6) hex = 'FF$hex';
  if (hex.length != 8) return null;
  final intValue = int.tryParse(hex, radix: 16);
  return intValue == null ? null : Color(intValue);
}

File: frontend/src/lib/category-visuals.ts
import type { LucideIcon } from "lucide-react";
import {
  ArrowLeftRight,
  Baby,
  Banknote,
  BookOpen,
  Bus,
  Car,
  CarTaxiFront,
  Coffee,
  Droplets,
  Dumbbell,
  Film,
  Fuel,
  Gamepad2,
  Gift,
  GraduationCap,
  HandHeart,
  Heart,
  Home,
  Landmark,
  Music,
  PawPrint,
  PiggyBank,
  Plane,
  Receipt,
  Scissors,
  Shield,
  Shirt,
  ShoppingBag,
  ShoppingCart,
  Smartphone,
  Sparkles,
  Stethoscope,
  Tag,
  TrendingUp,
  UtensilsCrossed,
  Wifi,
  Zap,
} from "lucide-react";

// Ported from mobile/lib/src/theme/category_visuals.dart so both clients
// render the same color + icon for a given category name.
const PALETTE = [
  "#ef4444",
  "#f97316",
  "#f59e0b",
  "#10b981",
  "#06b6d4",
  "#3b82f6",
  "#6366f1",
  "#8b5cf6",
  "#ec4899",
  "#14b8a6",
];

const ICON_KEYWORDS: Array<[string[], LucideIcon]> = [
  [["food", "restaurant", "lunch", "dinner", "meal"], UtensilsCrossed],
  [["grocer"], ShoppingCart],
  [["coffee", "cafe", "tea"], Coffee],
  [["transport", "commute"], Bus],
  [["uber", "taxi", "ride"], CarTaxiFront],
  [["fuel", "gas", "petrol"], Fuel],
  [["car", "vehicle"], Car],
  [["travel", "flight", "trip", "holiday"], Plane],
  [["rent", "home", "house"], Home],
  [["bill"], Receipt],
  [["utilit", "electric", "power"], Zap],
  [["water"], Droplets],
  [["internet", "broadband"], Wifi],
  [["phone", "mobile"], Smartphone],
  [["shop"], ShoppingBag],
  [["cloth", "fashion", "dress"], Shirt],
  [["health", "care"], Heart],
  [["medic", "pharma", "doctor", "hospital"], Stethoscope],
  [["gym", "fitness", "sport"], Dumbbell],
  [["education", "school", "tuition", "course"], GraduationCap],
  [["book"], BookOpen],
  [["entertain", "movie", "cinema"], Film],
  [["music"], Music],
  [["game"], Gamepad2],
  [["gift"], Gift],
  [["donation", "charity", "zakat"], HandHeart],
  [["salary", "wage", "payroll"], Banknote],
  [["income", "saving"], PiggyBank],
  [["invest", "stock", "dividend"], TrendingUp],
  [["tax", "bank", "loan", "emi"], Landmark],
  [["insurance"], Shield],
  [["pet"], PawPrint],
  [["baby", "kid", "child"], Baby],
  [["beauty", "spa"], Sparkles],
  [["salon", "hair", "barber"], Scissors],
  [["transfer"], ArrowLeftRight],
];

function hashName(name: string): number {
  let hash = 0;
  for (let i = 0; i < name.length; i += 1) {
    hash = (hash * 31 + name.charCodeAt(i)) >>> 0;
  }
  return hash;
}

export interface CategoryVisual {
  color: string;
  Icon: LucideIcon;
}

export function categoryVisual(
  name: string | null | undefined,
  storedColor?: string | null,
): CategoryVisual {
  const label = (name ?? "").trim();
  const lower = label.toLowerCase();

  const color =
    storedColor && /^#([0-9a-f]{6})$/i.test(storedColor)
      ? storedColor
      : PALETTE[label ? hashName(lower) % PALETTE.length : 9];

  let Icon: LucideIcon = Tag;
  for (const [keywords, candidate] of ICON_KEYWORDS) {
    if (keywords.some((keyword) => lower.includes(keyword))) {
      Icon = candidate;
      break;
    }
  }

  return { color, Icon };
}

// Report chart series palette — same order as the Flutter reports page.
export const CHART_COLORS = [
  "#0f766e",
  "#2563eb",
  "#f59e0b",
  "#dc2626",
  "#7c3aed",
  "#0891b2",
  "#16a34a",
  "#db2777",
];
